x-pack/legacy/plugins/lens/public/editor_frame_plugin/editor_frame/data_panel_wrapper.tsx

```diff
@@ -21,11 +22,11 @@ interface DataPanelWrapperProps {
 }
 
 export const DataPanelWrapper = memo((props: DataPanelWrapperProps) => {
-  const setDatasourceState = useMemo(
-    () => (newState: unknown) => {
+  const setDatasourceState: StateSetter<unknown> = useMemo(
+    () => updater => {
       props.dispatch({
         type: 'UPDATE_DATASOURCE_STATE',
-        newState,
+        updater,
```

Comment:
This doesn't seem right. Actions are supposed to be serializable / deserializable. It's one of the benefits of using an action / reducer pattern.

I'm not sure what we should do here, but I'd like to think it through a bit more. If we ever end up using a reducer-based undo / redo system, it would be really nice to keep actions data-only.

Reply:
You might be right in the world of redux, but I'm not sure that pattern applies in the build-in dispatcher.

Also, this concept is already something we merged for updating layers: https://github.com/elastic/kibana/blob/feature/lens/x-pack/legacy/plugins/lens/public/editor_frame_plugin/editor_frame/state_management.ts#L44

Reply:
It's also not a strict rule in redux itself (https://redux.js.org/faq/actions#why-should-type-be-a-string-or-at-least-serializable-why-should-my-action-types-be-constants), but you are right, we should be aware of the trade-offs. I thought about this and discussed with @streamich when doing this change to the layer state management and I think it's the right thing to do this here.

For the undo / redo system the easiest way to is to buffer the sequence of states in a separate property (like this: https://redux.js.org/recipes/implementing-undo-history) to be able to re-apply them. The actions don't have to be serializable for this to work because they are still only applied once. Another way out of this would be to keep the actions as they are for the `dispatch` calls and call the updater functions in a middleware instead of the actual reducer to keep it serializable (just like thunk and saga works). But I would only do that once we get to the point we actually need it. We should take the easiest way that doesn't paint us in a corner here IMHO.